pkg/controller/replicaschedulingpreference/controller.go

```diff
@@ -0,0 +1,751 @@
+	s.reviewDelay = 50 * time.Millisecond
+	s.smallDelay = 20 * time.Millisecond
+	s.updateTimeout = 5 * time.Second
+}
+
+func (s *ReplicaSchedulingPreferenceController) Run(stopChan <-chan struct{}) {
+	go s.controller.Run(stopChan)
+	go s.templateController.Run(stopChan)
+	go s.overrideController.Run(stopChan)
+	go s.placementController.Run(stopChan)
+
+	s.targetInformer.Start()
+	s.podInformer.Start()
+	s.deliverer.StartWithHandler(func(item *util.DelayingDelivererItem) {
+		s.workQueue.Add(item)
+	})
+	s.clusterDeliverer.StartWithHandler(func(_ *util.DelayingDelivererItem) {
+		s.reconcileOnClusterChange()
+	})
+
+	go wait.Until(s.worker, time.Second, stopChan)
+
+	util.StartBackoffGC(s.backoff, stopChan)
+
+	// Ensure all goroutines are cleaned up when the stop channel closes
+	go func() {
+		<-stopChan
+		s.targetInformer.Stop()
+		s.podInformer.Stop()
+		s.workQueue.ShutDown()
+		s.deliverer.Stop()
+		s.clusterDeliverer.Stop()
+	}()
+}
+
+type reconciliationStatus int
+
+const (
+	statusAllOK reconciliationStatus = iota
+	statusNeedsRecheck
+	statusError
+	statusNotSynced
+)
+
+func (s *ReplicaSchedulingPreferenceController) worker() {
+	for {
+		obj, quit := s.workQueue.Get()
+		if quit {
+			return
+		}
+
+		item := obj.(*util.DelayingDelivererItem)
+		qualifiedName := item.Value.(*util.QualifiedName)
+		status := s.reconcile(*qualifiedName)
+		s.workQueue.Done(item)
+
+		switch status {
+		case statusAllOK:
+			break
+		case statusError:
+			s.deliver(*qualifiedName, 0, true)
+		case statusNeedsRecheck:
+			s.deliver(*qualifiedName, s.reviewDelay, false)
+		case statusNotSynced:
+			s.deliver(*qualifiedName, s.clusterAvailableDelay, false)
+		}
+	}
+}
+
+func (s *ReplicaSchedulingPreferenceController) deliverObj(obj pkgruntime.Object, delay time.Duration, failed bool) {
+	// TODO: right now this works on the RSP being same "ns/name" as the
+	// deploymenttemplate. Update this to rsp being able to use
+	// targetRef (kind+name) with multiple kinds.
+	// For reviewers - Plan is to use Objectmeta.OwnerReferences by setting it to
+	// created rsp into the target object. This way when reconcile is called
+	// the information will be available to link objects both ways.
+	qualifiedName := util.NewQualifiedName(obj)
+	s.deliver(qualifiedName, delay, failed)
+}
+
+// Adds backoff to delay if this delivery is related to some failure. Resets backoff if there was no failure.
+func (s *ReplicaSchedulingPreferenceController) deliver(qualifiedName util.QualifiedName, delay time.Duration, failed bool) {
+	key := qualifiedName.String()
+	if failed {
+		s.backoff.Next(key, time.Now())
+		delay = delay + s.backoff.Get(key)
+	} else {
+		s.backoff.Reset(key)
+	}
+	s.deliverer.DeliverAfter(key, &qualifiedName, delay)
+}
+
+// Check whether all data stores are in sync. False is returned if any of the informer/stores is not yet
+// synced with the corresponding api server.
+func (s *ReplicaSchedulingPreferenceController) isSynced() bool {
```

Comment:
What about the sync status for pods, templates, overrides and placement?

Reply:
Updated!